cram_pr2/cram_pr2_projection/src/ik.lisp

```diff
@@ -154,48 +154,91 @@
-                                             (cram-tf:ensure-pose-in-frame
-                                              cartesian-pose
-                                              cram-tf:*robot-torso-frame*
-                                              :use-zero-time t))
-                (:joint_state :robot_state :ik_request) (or seed-state
-                                                            (make-zero-seed-state left-or-right))
-                (:timeout :ik_request) 1.0)))
+                (roslisp:call-service
+                 (concatenate 'string (getf *ik-service-namespaces* left-or-right) "/get_ik")
+                 "moveit_msgs/GetPositionIK"
+                 (roslisp:make-request
+                  "moveit_msgs/GetPositionIK"
+                  (:ik_link_name :ik_request) ik-link
+                  (:pose_stamped :ik_request) (cl-transforms-stamped:to-msg
+                                               (cram-tf:ensure-pose-in-frame
+                                                cartesian-pose
+                                                cram-tf:*robot-torso-frame*
+                                                :use-zero-time t))
+                  (:joint_state :robot_state :ik_request) (or seed-state
+                                                              (make-zero-seed-state left-or-right))
+                  (:timeout :ik_request) 1.0)))
           (cond ((eql response-error-code
                       (roslisp-msg-protocol:symbol-code
                        'moveit_msgs-msg:moveiterrorcodes
-                       :success)) joint-state)
+                       :success))
+                 (roslisp:set-debug-level NIL dbg-lvl)
+                 (values joint-state test-angle))
                 ((eql response-error-code
                       (roslisp-msg-protocol:symbol-code
                        'moveit_msgs-msg:moveiterrorcodes
                        :no_ik_solution)) nil)
                 (t (error 'simple-error
-                          :format-control "IK service failed: ~a"
-                          :format-arguments (list
-                                             (roslisp-msg-protocol:code-symbol
-                                              'moveit_msgs-msg:moveiterrorcodes
-                                              response-error-code))))))
+                                  :format-control "IK service failed: ~a"
+                                  :format-arguments (list
+                                                     (roslisp-msg-protocol:code-symbol
+                                                      'moveit_msgs-msg:moveiterrorcodes
+                                                      response-error-code))))))
       (simple-error (e)
         (declare (ignore e))
-        (format t "IK service call freaked out. No IK solution found.~%")))))
+        (cut:with-vars-bound (?upper-limit ?lower-limit)
```

Comment:
Lisp prolog calls are quite expensive in the current Prolog implementation, let's try to avoid unnecessary calls. So 2 solutions for this:

* implement a caching mechanism to not call prolog if limits are already known
* pass the limit as an argument to call_ik

I would prefer 2nd solution.

Reply:
Implemented the second one.